driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -77,28 +79,23 @@ impl UpdatingOrderbook {
         };
     }
 
-    /// Use the context, ensuring that the orderbook has been initialized.
-    fn do_with_context<T>(&self, callback: impl FnOnce(&mut Context) -> Result<T>) -> Result<T> {
-        // We unwrap because there is no way the mutex can be poisoned. If the mutex somehow was
-        // poisoned anyway against our expectation then panicking is correct because we cannot
-        // recover from this.
-        // The mutex cannot be poisoned because `do_with_context` is the only function accessing
-        // the mutex and is only called within this file. The `callback`s we use do not panic.
-        let mut context_guard = self.context.lock().expect("poison error");
+    /// Ensure that the orderbook has been initialized and updated.
+    /// Post condition on success: self.context is Some(_).
+    async fn update_context(&self) -> Result<()> {
```

Comment:
Is there a reason we no longer use `do_with_context`? I thought it was a nice abstraction. The function signature would be a bit clunky though:
```
async fn do_with_context<T, F, Fut>(&self, callback: F) -> Result<T>
where
    F: FnOnce(&mut Context) -> Fut,
    Fut: Future<Output = Result<T>>,
{
```

Reply:
**edit**: I see your comment now.

Reply:
Hmm, this is a tough one. It looks like if you declare a `Fut` generic as I did above, you get compile errors because `Fut` needs to live as long as the `do_with_context` call, but it can only live as long as `&mut Context` parameter to the function. Usually this is solved with higher rank trait lifetime bounds, but that only works for the `F` type :cry: 

One way to work around this is require the callback to be a boxed future:
```rust
    async fn do_with_context<T, F>(&self, callback: F) -> Result<T>
    where
        F: for<'ctx> FnOnce(&'ctx mut Context) -> BoxFuture<'ctx, Result<T>>,
    {
        let mut context_guard = self.context.lock().await;
        match context_guard.as_mut() {
            Some(context) => {
                self.update_with_events(context).await?;
                callback(context).await
            }
            None => {
                let mut context = Context {
                    orderbook: Orderbook::default(),
                    last_handled_block: 0,
                    block_timestamp_reader: CachedBlockTimestampReader::new(self.web3.clone()),
                };
                self.load_orderbook_from_file(&mut context);
                self.update_with_events(&mut context).await?;
                let result = callback(&mut context).await;
                *context_guard = Some(context);
                result
            }
        }
    }

```

Reply:
@e00E Not sure if you saw this comment

Reply:
Thanks. I am using this now.